### PR TITLE
Lra 467 mmss travel system for collecting students information

### PR DESCRIPTION
**Pull request by britaumich**
To use radio buttons in the travel form I had to upgrade Tailwind. To do that I run these commands:
```
yarn remove tailwindcss postcss autoprefixer
yarn add -D tailwindcss@npm:@tailwindcss/postcss7-compat postcss@^7 autoprefixer@^9
yarn add @tailwindcss/forms
```
Other changes are in tailwind.config.js and in application.sass.

Radio buttons are tricky to work with:

- To make a set of radio buttons required in a new form, 'required' should be added to the first button
- To make a correct button checked in the edit form I had to add radio_button_checked helper (in application_helper)

The workflow process:

- the travel form is based on this form https://docs.google.com/forms/d/1P84G5AYkCQgDmfn3Xzw4-Q6UeiasXd_z3HreAN6e680/edit
- if the application status is enrolled the user will see the "Add Travel Form" button. 
- All uploaded travel forms are displayed in the sidebox. 
- Multiple forms might be uploaded. 
- Applicants can edit their forms
-Admins can see applicants' travel forms in the active admin Application Show view. 
- Admins can see/add/edit/delete travel forms in the Active Admin Travel views.

Please read all travel instructions carefully.

### CODE REVIEW

File: app/views/travels/_form.html.erb [hunk 1]
```diff
@@ -1,57 +1,187 @@
...
+      All fields are required unless marked <span class="text-gray-700">(optional)</span>
+    </div>
+    <hr>
+    <%= form.label 'travel[arrival_session]', 'Session of Arrival *' %>
+    <div class="form-check">
+      <% @sessions.each_with_index do |s, index| %>
+        <div>
+          <% if index == 0 %>
+            <input class="form-check-input appearance-none rounded-full h-4 w-4 border border-gray-600 bg-white checked:bg-blue-600 checked:border-blue-600 focus:outline-none transition duration-200 mt-1 align-top bg-no-repeat bg-center bg-contain float-left mr-2 cursor-pointer" 
+            type="radio" name="travel[arrival_session]" id="<%= s %>" value="<%= s %>"
```
Comment (rsmoke) [inline app/views/travels/_form.html.erb:28]: I believe there is a way to deal with the radio buttons that is a Rails idiom

File: app/views/travels/_form.html.erb [hunk 2]
```diff
@@ -1,57 +1,187 @@
-<%= form_with(model: travel, local: true) do |form| %>
-  <% if travel.errors.any? %>
+<div>
```
Comment (rsmoke) [inline app/views/travels/_form.html.erb:1]: It seems that we can make the whole form easier to work with by checking if the traveller is using car or other means of transportation and showing fields that based on the selection

File: db/migrate/20221214214107_add_fields_to_travel.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class AddFieldsToTravel < ActiveRecord::Migration[6.1]
+  def change
+    rename_column :travels, :transport_needed, :arrival_transport
+    rename_column :travels, :carrier, :arrival_carrier
+    rename_column :travels, :route_num, :arrival_route_num
+    add_column :travels, :arrival_date, :date
+    add_column :travels, :arrival_time, :time
+    add_column :travels, :depart_transport, :string
+    add_column :travels, :depart_route_num, :time
```
Comment (rsmoke) [inline db/migrate/20221214214107_add_fields_to_travel.rb:9]: fixed in later migration

**Review by rsmoke — CHANGES_REQUESTED** [review]
see inline comments and remove the .old1 file

File: app/javascript/controllers/travel_controller.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import { Controller } from 'stimulus';
+export default class extends Controller {
+  static targets = ["depart", "arrival", "message"]
+
+  checkDates(event) {
+    let arrival = this.arrivalTarget.value;
+    let depart = this.departTarget.value;
+
+    if (typeof arrival !== 'undefined' && typeof arrival !== 'undefined') {
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/javascript/controllers/travel_controller.js]: fixed here [7f9c618](https://github.com/lsa-mis/mmss-mysql/pull/107/commits/7f9c6185debf5c77583315a8bec931ca9c381f4f)

File: db/migrate/20221214214107_add_fields_to_travel.rb (hunk 3 above)
Comment (britaumich) [reply db/migrate/20221214214107_add_fields_to_travel.rb:9]: had to run annotate, fixed here [aa61a19](https://github.com/lsa-mis/mmss-mysql/pull/107/commits/aa61a19dd8c7be6b40aefd5b2939a7a2c058621e)

File: app/views/static_pages/_directions.html.erb [hunk 5]
```diff
@@ -122,8 +122,21 @@
       <div class="m-2">
         <p class="direction_notes">
           We look forward to seeing you at the <%= @current_camp_year %> University of Michigan MMSS High School Summer Program.
+          <br><br>
+        <% if @current_enrollment.travels.present? %>
+          Please check the progress window at the right for you uploaded travel forms.
+          If you are attending more than one session and have more then one trip please fill a separate form for every trip.
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/views/static_pages/_directions.html.erb]: fixed here [71edf70](https://github.com/lsa-mis/mmss-mysql/pull/107/commits/71edf70d08af4cffd8bf09376956a6bd578aad7e)

File: app/views/static_pages/_directions.html.erb [hunk 6]
```diff
@@ -122,8 +122,21 @@
...
+        <% else %>
+          We would like to know your travel information so that we may plan for your arrival to, and departure from, the campus. 
+          Please fill the form below. 
+          If you are attending more than one session and have more then one trip please fill a separate form for every trip.
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/views/static_pages/_directions.html.erb]: fixed

File: app/admin/enrollments.rb [hunk 7]
```diff
@@ -314,6 +314,23 @@
...
+        table_for application.travels do
+          column "Session of Arrival" do |item|
+            item.arrival_session
+          end
+          column "Session of Departure" do |item|
+            item.depart_session
+          end
+          column "Travel Forms" do |item|
+            link_to("view", admin_travel_path(item))
+          end
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/admin/enrollments.rb:329]: done

File: app/models/camp_occurrence.rb [hunk 8]
```diff
@@ -55,4 +55,8 @@ def display_name
     "#{self.description} - #{self.begin_date} to #{self.end_date}" # or whatever column you want
   end
 
+  def description_with_start_day
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/models/camp_occurrence.rb]: I like to display sessions here like "Session 2: September 23 to September 30", 
not "Session 2: 2022-09-04 to 2022-10-10"

File: app/models/travel.rb [hunk 9]
```diff
@@ -2,18 +2,29 @@
...
+#  arrival_transport :string(255)
+#  arrival_carrier   :string(255)
+#  arrival_route_num :string(255)
+#  note              :text(65535)
+#  created_at        :datetime         not null
+#  updated_at        :datetime         not null
+#  arrival_date      :date
+#  arrival_time      :time
+#  depart_transport  :string(255)
+#  depart_route_num  :time
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/models/travel.rb]: You meant "time' - fixed

File: app/views/travels/_form.html.erb (hunk 2 above)
Comment (britaumich) [reply app/views/travels/_form.html.erb:1]: added functionality to hide and show transportation details depending on transportation type. Details are displayed for airplanes, trains, and buses.

**Comment by britaumich** [comment]
@rsmoke It would be really nice if you fix the app on your computer and check the workflow of this whole process. There are still questions to discuss but you have to see how it looks and works.

File: app/helpers/application_helper.rb [hunk 10]
```diff
@@ -144,4 +144,29 @@ def show_international(international)
     end
   end
 
+  def transportation
+    ['Airplane', 'Bus', 'Train', 
+      'Automobile - parent or permitted designee is driving me to the University of Michigan campus',
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/helpers/application_helper.rb:149]: I'll ask Doreen about that

**Comment by rsmoke** [comment]
Currently an Admin cannot create new travel. It throws an error:
```
ActiveRecord::RecordNotFound at /admin/travels/new
Couldn't find Enrollment without an ID
```

**Comment by rsmoke** [comment]
The sidebar direction for travel needs to populate when the status moves to enrolled. Currently there is no notification in the sidebar untila travel gets created

**Comment by rsmoke** [comment]
In adminUI under the travel panel maybe populate more information like Transport type, Time of Arrival and Departure, and Airline/Flight number

File: app/helpers/application_helper.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (britaumich) [reply app/helpers/application_helper.rb:149]: Doreen's reply: "Automobile means they are NOT commuters but are arriving via a car that does not need our support but we would like them to respond so we know how they are arriving.  No other details are needed for the option.
NA means commuter so perhaps we adjust that to say commuter?"
This is the list of transportation types:

- Airplane
- Bus
- Train
- Automobile - parent or permitted designee is driving me to the University of Michigan campus
- I am a daily MMSS commuter

This is what we have now.

**Comment by britaumich** [comment]
Active admin edit form works. 
The new form works but don't check validations.

**Comment by britaumich** [comment]
> The sidebar direction for travel needs to populate when the status moves to enrolled. Currently there is no notification in the sidebar until travel gets created

It works the same way as applicant_detail, session, and recommendation. There is a button to Add Travel form in directions.

**Comment by britaumich** [comment]
@rsmoke 
Now the travel form (on users site) validates (by stimulus controller travel_controller.js) the following rules:
```
If transportation is 'airplane', 'bus' or 'train', we need
career name
flight/bus/train/ number
date
time

If transportation is "automobile', we need
date
time

If transportation is 'commuter' we don't need any other details.
```

There is no such validation for active admin travel forms.

The travel New form still needs better styles.

File: app/views/travels/_form.html.erb (hunk 1 above)
Comment (britaumich) [reply app/views/travels/_form.html.erb]: I replaced radio buttons with drop-down lists

File: app/views/static_pages/_directions.html.erb [hunk 11]
```diff
@@ -122,8 +122,21 @@
       <div class="m-2">
         <p class="direction_notes">
           We look forward to seeing you at the <%= @current_camp_year %> University of Michigan MMSS High School Summer Program.
+          <br><br>
+        <% if @current_enrollment.travels.present? %>
+          Please check the progress window at the right for you uploaded travel forms.
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/views/static_pages/_directions.html.erb]: fixed

File: app/helpers/application_helper.rb [hunk 12]
```diff
@@ -144,4 +144,29 @@ def show_international(international)
...
+
+  def show_date(field)
+    field.strftime("%A, %d %b %Y") unless field.blank?
+  end
+
+  def show_time(field)
+    field.strftime("%I:%M %p") unless field.blank?
+  end
+
+  def radio_button_checked(object, key, value)
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/helpers/application_helper.rb]: There are no radio buttons anymore

**Comment by britaumich** [comment]
Added more styles.
Please test everything again.

**Review by rsmoke — CHANGES_REQUESTED** [review]
Other than my inline Q about using let vs var in the travel_controller.js this looks good. Let me know your answer to the question.

**Review by rsmoke — APPROVED** [review]

**Pull request merged**